### PR TITLE
`LightningVJPs` works with new `LightningQubit`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Lots of test cases were commented out once `LightningQubit` started using the new device API due to not having VJP support. This PR adds the needed changes for the `LightningQubit.compute_vjp` method to be fully integrated.

**Description of the Change:**
* Add logic to `LightningVJPs` to correctly handle VJP computation with devices using the new API.
* Uncomment interface tests using `LightningQubit` with `device_vjp=True`.
* Skip/xfail some unsupported tests. This mostly includes tests for state adjoint differentiation, which is not supported on `LightningQubit`.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
This PR should not get merged until Lightning PR https://github.com/PennyLaneAI/pennylane-lightning/pull/644 is merged.

**Comment by mudit2812** [comment]
[sc-60397]

File: pennylane/workflow/jacobian_products.py
Comment (mudit2812) [inline pennylane/workflow/jacobian_products.py:752]: The `pragma: no cover` was adding to this line because the branch in question is used for old API devices (`lightning.kokkos` and `lightning.gpu`), which are not included in the testing suite.

File: pennylane/workflow/jacobian_products.py
Comment (dwierichs) [inline pennylane/workflow/jacobian_products.py:751]: I understand that we need to separate out the execution kwarg if it is `None`, in order to not override the `DefaultExecutionConfig` in `_device.compute_vjp`. However, I'm wondering whether 
1. this is the most stable API for `compute_vjp`, or whether maybe the `DefaultExecutionConfig` should be passed down by default, confirming the default instead of creating a separate logical branch for it, and
2. there is a reason that we do not to do this for DQ? :thinking: Maybe I'm missing it and we actually are doing it, but I did not find it.

File: tests/devices/test_lightning_qubit.py
Comment (dwierichs) [inline tests/devices/test_lightning_qubit.py:92]: Are the changes in this file related to the PR immediately, or more of a byproduct?

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: This is not optimal, I suppose? Shouldn't the QNode return value be independent of the device? :thinking: Sorry if this is an established behaviour, I may just not be used to it.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: This is not necessary, I think, because below all tests with `diff_method!="backprop"` are `skip`ped anyways. If we want to switch to `xfail`, I would recommend that we do it for all devices.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1367]: As for Autograd.

File: tests/interfaces/default_qubit_2_integration/test_jax_qnode_default_qubit_2.py
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_qnode_default_qubit_2.py:1246]: As for Autograd

File: tests/interfaces/default_qubit_2_integration/test_torch_qnode_default_qubit_2.py
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_torch_qnode_default_qubit_2.py:1073]: As for Autograd

File: tests/interfaces/test_jacobian_products.py
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:72]: :tada:

**Review by dwierichs — CHANGES_REQUESTED** [review]
Thanks @mudit2812, mostly looks good to me!
Just had some questions regarding the modified dtype test and the logic of the default execution config, as well as the strategy for introducing `xfail`s in some test.

Are we not supporting Tensorflow in general with Lightning device VJPs? If so, why is that (might be an ignorant question, sorry).

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: If `adjoint` diff is requested, the device executes and computes gradients in a single call rather than separate ones. So, this test fails with lightning before it reaches the `pytest.skip` line. The reason why only lightning tests are skipped is because `default.qubit` supports all state-based measurements with adjoint diff now.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: Good question. @albi3ro and I have discussed this briefly offline, and we decided it was okay to have `float` as a valid dtype for scalar returns with no interface.

Comment (dwierichs) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:99]: Fair enough! I hope this does not break any further processing, but as lightning got its own VJP method, I suppose it is fine :) Is this documented anywhere? :)

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (dwierichs) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:1323]: Oh, I see! So it is not about the differentiation being supported, but about the forward pass with a specific diff_method "in mind" being supported? Thanks! :)

File: pennylane/workflow/jacobian_products.py
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:724]: This is regular python syntax for objects not yet initialized. In this case, it is because of circular imports.

File: pennylane/workflow/jacobian_products.py
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:751]: The `DeviceJacobianProducts` class only works with devices using the new API, hence includes the `ExecutionConfig` as a required positional argument. However, `LightningVJPs` is supposed to work with both the old and new device API, since LGPU and LKokkos still use the old API, which does not work with `ExecutionConfig`s. I think it makes more sense to leave this as is for now, and then we can remove these branches, or maybe even this class altogether once LGPU and LKokkos are updated as well.

**Review by albi3ro — CHANGES_REQUESTED** [review]
`lightning.qubit` that implements the new device API should strictly use `DeviceJacobianProducts` and `DeviceJacobians`.  

`LightningVJPs` is strictly a patch to handle the vjp interface that was unique to the legacy lightning interface, and should now only apply to `lightning.kokkos` and `lightning.gpu`.

File: tests/devices/test_lightning_qubit.py
Comment (mudit2812) [reply tests/devices/test_lightning_qubit.py:92]: It's a byproduct of changing `LightningQubit` to use the new API. Somewhere along processing measurements, the dtype gets converted to double precision. This needs to be investigated, but that's a task for another PR.

File: pennylane/workflow/jacobian_products.py
Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:751]: Had some discussion with @albi3ro offline. Execution has been updated to use `DeviceJacobianProducts` with `lightning.qubit`.

File: tests/interfaces/test_jacobian_products.py
Comment (albi3ro) [inline tests/interfaces/test_jacobian_products.py:229]:
```suggestion
            pytest.xfail("Lightning devices don't have JVP method")
```
? Maybe.  Can't hurt and will let us know if things change.

**Review by dwierichs — APPROVED** [review]
Nice, @mudit2812 !

**Review by vincentmr — APPROVED** [review]
Thanks @mudit2812 , good to go on my side too.

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:538]:
```suggestion
        and "lightning" not in getattr(device, "short_name", "").lower()
```
This is now for qpu and kokkos which still implement the old device interface.

Comment (albi3ro) [reply pennylane/workflow/execution.py:538]: Though... I'm confused because we should still support VJPs for them through the `LightningVJPs` class?

Comment (mudit2812) [reply pennylane/workflow/execution.py:538]: I chose `name` because LGPU and LK have a `name` attribute along the lines of `"Lightning ... plugin for PennyLane"`, so it still works, and we won't have to change this again once they're ported to the new API.

Comment (mudit2812) [reply pennylane/workflow/execution.py:538]: > Though... I'm confused because we should still support VJPs for them through the LightningVJPs class?

This condition is for raising an error if we're requesting `device_vjp=True` with an old API device, and we don't want to include lightning devices in that list.

File: pennylane/workflow/jacobian_products.py
Comment (albi3ro) [inline pennylane/workflow/jacobian_products.py:688]: Could we instead make this a constraint on `Device.compute_vjp`?  I think it it reasonable to add the constraint that the pytree structure of a VJP should match the structure of `tape.get_parameters()`.

Comment (mudit2812) [inline pennylane/workflow/jacobian_products.py:688]: I'm not particularly fond of this as it's doing the job of the device. We've specified the return type specification quite clearly, and devices should obey that. I've added this for now for the sake of getting the tests to pass, but long term I think it's better to update the Lightning device(s) to fix the return type.

Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:688]: Talked offline. We will leave this as is for now, update lightning to correctly handle vjp return types, and then remove this logic after that PR is merged.

Comment (mudit2812) [reply pennylane/workflow/jacobian_products.py:688]: Yep, I'll leave it like this for now, and we can update `LightningQubit` accordingly and revert this after that change is merged.

File: tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py
Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:104]: Not currently. I will discuss with core team offline and update docs if we decide to.

Comment (mudit2812) [reply tests/interfaces/default_qubit_2_integration/test_autograd_qnode_default_qubit_2.py:104]: @albi3ro  has a doc PR that will address this.

**Review by albi3ro — APPROVED** [review]
Thanks for the changes 👍

**Pull request merged**
